lib/Conversion/DCToHW/DCToHW.cpp

```diff
@@ -0,0 +1,946 @@
+  return TypeSwitch<Type, Value>(type)
+      .Case<NoneType>([&](NoneType) { return rtlb.constant(0, 0); })
+      .Case<IntType, IntegerType>([&](auto type) {
+        return rtlb.constant(type.getIntOrFloatBitWidth(), 0);
+      })
+      .Case<hw::StructType>([&](auto structType) {
+        SmallVector<Value> zeroValues;
+        for (auto field : structType.getElements())
+          zeroValues.push_back(createZeroDataConst(rtlb, loc, field.type));
+        return rtlb.b.create<hw::StructCreateOp>(loc, structType, zeroValues);
+      })
+      .Default([&](Type) -> Value {
+        emitError(loc) << "unsupported type for zero value: " << type;
+        assert(false);
+        return {};
+      });
+}
+
+static bool isZeroWidthType(Type type) {
+  if (auto intType = type.dyn_cast<IntegerType>())
+    return intType.getWidth() == 0;
+  return type.isa<NoneType>();
+}
+
+static UnwrappedIO unwrapIO(Location loc, ValueRange operands,
+                            TypeRange results,
+                            ConversionPatternRewriter &rewriter,
+                            BackedgeBuilder &bb) {
+  RTLBuilder rtlb(loc, rewriter);
+  UnwrappedIO unwrapped;
+  for (auto in : operands) {
+    assert(isa<esi::ChannelType>(in.getType()));
+    InputHandshake hs;
+    auto ready = std::make_shared<Backedge>(bb.get(rtlb.b.getI1Type()));
+    auto [data, valid] = rtlb.unwrap(in, *ready);
+    hs.valid = valid;
+    hs.ready = ready;
+    hs.data = data;
+    hs.channel = in;
+    unwrapped.inputs.push_back(hs);
+  }
+  for (auto outputType : results) {
+    outputType = toESIHWType(outputType);
+    esi::ChannelType channelType = cast<esi::ChannelType>(outputType);
+    OutputHandshake hs;
+    Type innerType = channelType.getInner();
+    Value data;
+    if (isZeroWidthType(innerType)) {
+      // Feed the ESI wrap with an i0 constant.
+      data =
+          rewriter.create<hw::ConstantOp>(loc, rewriter.getIntegerType(0), 0);
+    } else {
+      // Create a backedge for the unresolved data.
+      auto dataBackedge = std::make_shared<Backedge>(bb.get(innerType));
+      hs.data = dataBackedge;
+      data = *dataBackedge;
+    }
+    auto valid = std::make_shared<Backedge>(bb.get(rewriter.getI1Type()));
+    auto [dataCh, ready] = rtlb.wrap(data, *valid);
+    hs.valid = valid;
+    hs.ready = ready;
+    hs.channel = dataCh;
+    unwrapped.outputs.push_back(hs);
+  }
+  return unwrapped;
+}
+
+static UnwrappedIO unwrapIO(Operation *op, ValueRange operands,
+                            ConversionPatternRewriter &rewriter,
+                            BackedgeBuilder &bb) {
+  return unwrapIO(op->getLoc(), operands, op->getResultTypes(), rewriter, bb);
+}
+
+///  Locate the clock and reset values from the parent operation based on
+///  attributes assigned to the arguments.
+static FailureOr<std::pair<Value, Value>> getClockAndReset(Operation *op) {
+  auto *parent = op->getParentOp();
```

Comment:
Could this be `getParentOpOfType<FuncOpIface>`? Or do you specifically want to restrict `dc.clock` to be in the same block args?